### PR TITLE
💾Added backend logic for Infractions & Reports, Moved User/Login Logic to the Core

### PR DESCRIPTION
**Pull request by sleeping-dev**
Added the backend logic necessary for #23.
This should give us all we need to create, edit and delete reports/infractions.

A few notes on the Put-Method for Reports:
I figured that once it is posted, the only property in a report that should be edited is the 'Dismiss'-Property, by mods. Which is why the Request Object only has this property settable, since a report shouldn't be changed after being submitted.

Infractions, on the other hand might be edited by mods if they accidentally banned a user for too long, or if the ban reason was inappropriate and another mod/admin changes it, etc...

@Breadkenty As for the frontend, you can refer to the request objects of the API controllers as to what data you need to send in the request body. This is of course up for debate, if we want to make changes here that's easily possible, I just designed the Request-Objects in a way that makes sense to me, but let me know if I missed anything or if you want any changes!

@skittlz444 I ran into the 'issue' that handling the responses from the core in the API controllers got a bit weird, since I wanted to move as much of the logic to the core as possible. You'll see I resorted to checking for null values/introducing a 'Success'-Property to the response object. To make things easier here I thought about introducing a 'ResponseBase' class that responses inherit from, together with a 'ResponseStatus' that helps describe the result of the Core's operations a bit more. I haven't actually used the class yet, I wanted to get your opinion first. What do you think about this? Is this a good approach?

I'll get to work on moving the rest to the Core when I find some time Sunday/Monday, and when I find time I'll revamp the Combo-Methods that are already in the core as well, there's some properties in the request-objects and dto's that we don't necessarely need.

### CODE REVIEW

**Comment by sleeping-dev** [comment]
I'm also in the process of refactoring the User-API logic to include reports/infractions when loading users, should be done soon.

**Comment by sleeping-dev** [comment]
Alright, this should take care of everything we need for moderation. Let me know what you think!

File: Smash_Combos.Core/Cqrs/Infractions/PutInfraction/PutInfractionRequest.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+namespace Smash_Combos.Core.Cqrs.Infractions.PutInfraction
+{
+    public class PutInfractionRequest : IRequest<PutInfractionResponse>
+    {
+        public int Id { get; set; }
+        public int? BanDuration { get; set; }
+        public int? Points { get; set; }
+        public InfractionCategory Category { get; set; }
+        public string Body { get; set; }
+        public bool LiftBan { get; set; }
```
Comment (Breadkenty) [inline Smash_Combos.Core/Cqrs/Infractions/PutInfraction/PutInfractionRequest.cs:16]: How does LiftBan work? How do we pass this request on the frontend? Correct me if I’m wrong but doesn’t LiftBan need to be in the Domain on the `Infraction` model to be able to pass it as `true` through the controller?

Should LiftBan be set to `private set` to prevent any users from doing some sneaky API requests to unban themselves?

Comment (sleeping-dev) [reply Smash_Combos.Core/Cqrs/Infractions/PutInfraction/PutInfractionRequest.cs:16]: We don't need to have LiftBan in the `Infraction-Model`, we only use the property on the `PutInfractionRequest`-Object that we expect in the `PutInfraction`-Method in the Controller.
You could think of it this way:
We no longer send the _object to change_ with the http-Request, we just send _the data we want changed_. The `PutInfractionRequest` contains the data that we want changed (Category, Body, Points, etc.) and a property to show if we want to lift the ban. 

In the `PutInfractionRequestHandler`-Class we can see how `LiftBan` works:
```
infraction.BanDuration = request.BanDuration;
infraction.Points = DeterminePoints(request);
infraction.Category = request.Category;
infraction.Body = request.Body;

if (request.LiftBan)
      infraction.BanLiftDate = DateTime.Now;
```
Basically the liftDate just gets set to the current time if the `LiftBan`-Property is true! There's no need to add it to the `Infraction`-class.

As for security, making it `private set` won't help too much (I'll still do it for good measure), as I think the biggest risk would be users _intercepting_ the http request before it gets to the server and changing the request-body, which wouldn't be stopped by the `private set`. I propose the following:
When a PutInfractionRequest comes in, we get the _current user ID_ and check if the currently logged in user is a Moderator or an Admin, and if he isn't, we deny the request. That way, even if the request is tampered with, as long as the user is not a mod or admin, nothing will happen.

**Comment by sleeping-dev** [comment]
One thing I just realized is missing is updating the SessionsController and moving that logic to the core. Since we want users to actually be denied a login when they're banned, haha. Will get to this after work.

**Comment by Breadkenty** [comment]
I think this all looks good, great work! My turn!
